File: lib/src/services/translation/providers/mymemory_provider.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../translation_models.dart';
import '../translation_provider.dart';

/// MyMemory translation API — keyless, supports en<->uk and many pairs.
///
/// `GET https://api.mymemory.translated.net/get?q=..&langpair=en|uk&de=email`
/// An email (optional) raises the anonymous quota from 5k to 50k chars/day.
/// Per-request text is capped at ~500 bytes, which is fine for words/phrases.
class MyMemoryProvider implements TranslationProvider {
  MyMemoryProvider(this._client, {this.email = ''});

  final http.Client _client;
  final String email;

  @override
  String get id => 'mymemory';

  @override
  String get name => 'MyMemory';

  @override
  bool get requiresConfiguration => false;

  @override
  Future<TranslationResult> translate({
    required String text,
    required String from,
    required String to,
  }) async {
    final uri = Uri.https('api.mymemory.translated.net', '/get', {
      'q': text,
      'langpair': '$from|$to',
      if (email.trim().isNotEmpty) 'de': email.trim(),
    });
    final http.Response resp;
    try {
      resp = await _client.get(uri);
    } catch (e) {
      throw ProviderException('Network error contacting MyMemory: $e');
    }
    if (resp.statusCode != 200) {
      throw ProviderException('MyMemory returned HTTP ${resp.statusCode}');
    }
    final json =
        jsonDecode(utf8.decode(resp.bodyBytes)) as Map<String, Object?>;
    final status = json['responseStatus'];
    final statusOk = status == 200 || status == '200';
    final data = json['responseData'] as Map<String, Object?>?;
    final primary = (data?['translatedText'] as String?)?.trim() ?? '';
    if (!statusOk || primary.isEmpty) {
      final detail = json['responseDetails']?.toString() ?? 'no translation';
      throw ProviderException('MyMemory: $detail');
    }

    final alternatives = <String>[];
    final matches = json['matches'];
    if (matches is List) {
      for (final m in matches) {
        if (m is Map) {
          final t = (m['translation'] as String?)?.trim();
          if (t != null &&
              t.isNotEmpty &&
              t.toLowerCase() != primary.toLowerCase() &&
              !alternatives.contains(t)) {
            alternatives.add(t);
          }
        }
      }
    }

    return TranslationResult(
      translatedText: primary,
      alternatives: alternatives.take(6).toList(),
      providerId: id,
      attribution: 'Translation by MyMemory',
    );
  }
}

File: ios-native/Eng/Services/Providers/MyMemoryProvider.swift
import Foundation

/// MyMemory translation API — keyless, supports en<->uk and many pairs.
/// `GET https://api.mymemory.translated.net/get?q=..&langpair=en|uk&de=email`
/// An email (optional) raises the anonymous quota from 5k to 50k chars/day.
struct MyMemoryProvider: TranslationProvider {
    let email: String
    var id: String { "mymemory" }

    func translate(text: String, from: String, to: String) async throws -> TranslationResult {
        var comps = URLComponents(string: "https://api.mymemory.translated.net/get")!
        comps.queryItems = [
            URLQueryItem(name: "q", value: text),
            URLQueryItem(name: "langpair", value: "\(from)|\(to)"),
        ]
        let trimmedEmail = email.trimmingCharacters(in: .whitespaces)
        if !trimmedEmail.isEmpty { comps.queryItems!.append(URLQueryItem(name: "de", value: trimmedEmail)) }

        let (data, resp): (Data, HTTPURLResponse)
        do { (data, resp) = try await Http.get(comps.url!) }
        catch { throw ProviderError("Network error contacting MyMemory: \(error.localizedDescription)") }
        guard resp.statusCode == 200 else { throw ProviderError("MyMemory returned HTTP \(resp.statusCode)") }

        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw ProviderError("MyMemory: bad response")
        }
        let status = json["responseStatus"]
        let statusOk = (status as? Int == 200) || (status as? String == "200")
        let responseData = json["responseData"] as? [String: Any]
        let primary = (responseData?["translatedText"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        if !statusOk || primary.isEmpty {
            let detail = (json["responseDetails"] as? String) ?? "no translation"
            throw ProviderError("MyMemory: \(detail)")
        }

        var alternatives: [String] = []
        if let matches = json["matches"] as? [[String: Any]] {
            for m in matches {
                if let t = (m["translation"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines),
                   !t.isEmpty, t.lowercased() != primary.lowercased(), !alternatives.contains(t) {
                    alternatives.append(t)
                }
            }
        }
        return TranslationResult(translatedText: primary,
                                 alternatives: Array(alternatives.prefix(6)),
                                 providerId: id,
                                 attribution: "Translation by MyMemory")
    }
}
